quimb/tensor/optimize.py

```diff
@@ -147,6 +156,28 @@ def parse_network_to_backend(tn, tags, constant_tags, to_constant):
         if isinstance(data, qarray):
             data = data.A
 
+        # if shared_tags are specified and this tensor contains one of them
+        if shared_tags and (t.tags & shared_tags):
+            if len(t.tags & shared_tags)>1:
```

Comment:
The logic is getting a bit complex with all the various tags, I wonder if its worth doing this way (pseudocode-ish):
```python
if not opt_tags:
    # handle simple case of every tensor being a variable
    ...
    return

for t in tn.select(tags, 'any'):
    # handle normally tagged tensors
    ...

for tag in shared_tags:
    for t in tn.select(tag):
        # handle shared tags
        ...
```
but let me know if that seems ambitious for you to implement - I can always push some changes too.